src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -1119,36 +1119,77 @@ PeerImp::onMessage(std::shared_ptr <protocol::TMShardInfo> const& m)
     // Parse the shard indexes received in the shard info
     RangeSet<std::uint32_t> shardIndexes;
     {
-        std::vector<std::string> tokens;
-        boost::split(tokens, m->shardindexes(), boost::algorithm::is_any_of(","));
-        for (auto const& t : tokens)
+        std::uint32_t earliestShard;
+        boost::optional<std::uint32_t> latestShard;
         {
-            std::vector<std::string> seqs;
-            boost::split(seqs, t, boost::algorithm::is_any_of("-"));
-            if (seqs.empty() || seqs.size() > 2)
+            auto const curLedgerSeq {
+                app_.getLedgerMaster().getCurrentLedgerIndex()};
+            if (auto shardStore = app_.getShardStore())
             {
-                fee_ = Resource::feeBadData;
-                return;
+                earliestShard = shardStore->earliestShardIndex();
+                if (curLedgerSeq >= shardStore->earliestSeq())
+                    latestShard = shardStore->seqToShardIndex(curLedgerSeq);
+            }
+            else
+            {
+                auto const earliestSeq {app_.getNodeStore().earliestSeq()};
+                earliestShard = NodeStore::seqToShardIndex(earliestSeq);
+                if (curLedgerSeq >= earliestSeq)
+                    latestShard = NodeStore::seqToShardIndex(curLedgerSeq);
             }
+        }
 
-            std::uint32_t first;
-            if (!beast::lexicalCastChecked(first, seqs.front()))
+        auto getIndex = [&](std::string const& s)
+            -> boost::optional<std::uint32_t>
+        {
+            std::uint32_t shardIndex;
+            if (!beast::lexicalCastChecked(shardIndex, s))
             {
                 fee_ = Resource::feeBadData;
-                return;
+                return boost::none;
+            }
+            if (shardIndex < earliestShard ||
+                (latestShard && shardIndex > latestShard))
+            {
+                fee_ = Resource::feeBadData;
+                JLOG(p_journal_.error()) <<
+                    "Invalid shard index " << shardIndex;
+                return boost::none;
             }
+            return shardIndex;
+        };
 
-            if (seqs.size() == 1)
-                shardIndexes.insert(first);
-            else
+        std::vector<std::string> tokens;
+        boost::split(tokens, m->shardindexes(),
+            boost::algorithm::is_any_of(","));
+        for (auto const& t : tokens)
+        {
+            std::vector<std::string> indexes;
```

Comment:
This can be made a bit more efficient by moving `indexes` outside of the loop and calling `indexes.clear()` at this spot.  It won't make any difference on the first iteration, but subsequent iterations can benefit from `indexes`'s `capacity()` to avoid allocations.